internal/provider/resources/resource_team_roles.go

```diff
@@ -82,15 +87,60 @@ func (r *teamRolesResource) MutateRoles(
 	teamId := data.TeamId.ValueString()
 
 	// Then convert the models to the request types for the API
-	workspaceRoles, diags := RequestWorkspaceRoles(ctx, data.WorkspaceRoles)
+	workspaceRoles, diags := common.RequestWorkspaceRoles(ctx, data.WorkspaceRoles)
 	if diags.HasError() {
 		return diags
 	}
-	deploymentRoles, diags := RequestDeploymentRoles(ctx, data.DeploymentRoles)
+	deploymentRoles, diags := common.RequestDeploymentRoles(ctx, data.DeploymentRoles)
 	if diags.HasError() {
 		return diags
 	}
 
+	// Check if deployment roles have corresponding workspace roles
+	// get list of deployment ids
+	deploymentIds := lo.Map(deploymentRoles, func(role iam.DeploymentRole, _ int) string {
+		return role.DeploymentId
+	})
+
+	// get list of deployments
+	listDeployments, err := r.platformClient.ListDeploymentsWithResponse(ctx, r.organizationId, &platform.ListDeploymentsParams{
+		DeploymentIds: &deploymentIds,
+	})
+	if err != nil {
+		tflog.Error(ctx, "failed to mutate Team roles", map[string]interface{}{"error": err})
+		diags.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to mutate Team roles and list deployments, got error: %s", err),
+		)
+		return diags
+	}
+	_, diagnostic := clients.NormalizeAPIError(ctx, listDeployments.HTTPResponse, listDeployments.Body)
+	if diagnostic != nil {
+		diags.Append(diagnostic)
+		return diags
+	}
+
+	// get list of workspace ids from deployments
+	deploymentWorkspaceIds := lo.Map(listDeployments.JSON200.Deployments, func(deployment platform.Deployment, _ int) string {
+		return deployment.WorkspaceId
+	})
+
+	// get list of workspaceIds
+	workspaceIds := lo.Map(workspaceRoles, func(role iam.WorkspaceRole, _ int) string {
+		return role.WorkspaceId
+	})
+
+	// check if deploymentWorkspaceIds are in workspaceIds
+	workspaceIds = lo.Intersect(lo.Uniq(workspaceIds), lo.Uniq(deploymentWorkspaceIds))
+	if len(workspaceIds) != len(deploymentWorkspaceIds) {
+		tflog.Error(ctx, "failed to mutate Team roles", map[string]interface{}{"error": err})
+		diags.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to mutate Team roles, not every deployment role has a corresponding workspace role"),
+		)
+		return diags
+	}
```

Comment:
can we make this into a helper so we can reuse in `resource_team_roels` and `resource_team`